### DISCUSSION TITLE
Is there any way to avoid the need to add the following lines to `config/services.php` for autoload/autowire?

### DISCUSSION THREAD

**Discussion by reypm** [symfony/symfony#42453] [Q&A] [answered]
I have a project with a few microservices and shared git submodules. Because of this setup I need to add the following lines to `config/services.php` on each microservice when needed:

```php
    $services->load('Apps\\Apps\\', __DIR__.'/../../Apps/Apps/*{Repository}.php');
    $services->load('Apps\\Apps\\', __DIR__.'/../../Apps/Apps/*{Interface}.php');

    $services->load('ElectronicDocument\\ElectronicDocument\\', __DIR__.'/../../ElectronicDocument/ElectronicDocument/*{Repository}.php');
    $services->load('ElectronicDocument\\ElectronicDocument\\', __DIR__.'/../../ElectronicDocument/ElectronicDocument/*{Interface}.php');

    $services->load('User\\Customer\\', __DIR__.'/../../User/Customer/*{Repository}.php');
    $services->load('User\\Customer\\', __DIR__.'/../../User/Customer/*{Interface}.php');
```

wonder if I can add something in the `Kernel.php` like a Compiler Pass or so for autoload/autoregister this as services. 🤔 

Any advice? How would you do it?

**Comment by ro0NL** [comment]
a bundle?

**Reply by reypm** [reply]
what do you mean by a bundle? sorry but I am not following you :(

**Reply by ro0NL** [reply]
or you could merge such default config into the kernel yes, either directly https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/5.4/src/Kernel.php#L14, or indirectly using a pass (or bundle :))

**Reply by ro0NL** [reply]
a bundle allows you to share logic basically: https://symfony.com/doc/current/bundles.html

**Reply by monteiro** [reply]
We use a bundle to all microservices install that bundle to have all services registered. Great course about bundles on symfonycasts: https://symfonycasts.com/screencast/symfony-bundle (learned everything that I know from that course).

**Reply by reypm** [reply]
I don't think a bundle would be good for us, lemme see if I can made it to work from the Kernel

**Reply by ro0NL** [reply]
why not? Have you read https://symfony.com/doc/current/bundles/extension.html

**Reply by ro0NL** [reply]
maybe a bundle is overkill. Semantically it allows you to share logic with any symfony kernel, thus any project.

If you patch the kernel, which is faster to do, the logic is tied to that specific kernel.

**Reply by reypm** [reply]
you're correct there, let me explore the bundle way with my co-workers I can't say yes or no by myself, thanks for now to both of you

**Comment by ahundiak** [comment]
I suppose you could write a services.php file for each of your namespaces.  Your config/services.php would then import the individual services.php file.  I don't know if you would really accomplish much though.  It might be a bit easier to maintain as the number of services in each namespace grows.

**Reply by reypm** [reply]
@ahundiak tried your solution but it does not work or I am missing something. Here is what I have done, lemme know if I am missing something or if you meant something different.

```php
<?php

declare(strict_types=1);

namespace App;

use Symfony\Bundle\FrameworkBundle\Kernel\MicroKernelTrait;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Symfony\Component\HttpKernel\Kernel as BaseKernel;
use Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator;
use function dirname;

class Kernel extends BaseKernel
{
    use MicroKernelTrait;

    private const MODULES = [
        __DIR__.'/../../Apps/Apps/',
        __DIR__.'/../../ElectronicDocument/ElectronicDocument/',
        __DIR__.'/../../Subscription/Subscription/',
        __DIR__.'/../../User/Customer/',
    ];

    protected function configureContainer(ContainerConfigurator $container): void
    {
        // ...

        foreach (self::MODULES as $modulePath) {
            $container->import($modulePath.'/config/module_config.php');
        }

        if (is_file(dirname(__DIR__).'/config/services.php')) {
            $container->import('../config/services.php');
            $container->import('../config/{services}_'.$this->environment.'.php');
        } elseif (is_file($path = dirname(__DIR__).'/config/services.php')) {
            (require $path)($container->withPath($path), $this);
        }
    }

    // ...
}
```

The config files looks like:
```php
<?php

declare(strict_types=1);

use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $services = $containerConfigurator->services();

    $services->load('ElectronicDocument\\ElectronicDocument\\', __DIR__.'/*{Repository}.php');
    $services->load('ElectronicDocument\\ElectronicDocument\\', __DIR__.'/*{Interface}.php');
};
```
the only changes between them are the paths used as namespaces. However does not seems to be working since I am getting the following error:

> Uncaught Exception: The service "App\Security\SecurityFactory" has a dependency on a non-existent service "Subscription\Subscription\UserSubscriptionRepositoryInterface".

which means the above code didn't autoload anything. Any ideas?

**Reply by ahundiak** [reply]
Pretty sure you need to set the auto defaults in each individual file:
```
$services->defaults()->autowire()->autoconfigure();
```
If that does not help then I'd really need a little test case.  I personally don't use the load method from individual services.php files.  I rather do something like:
```
    $services->set(PageTemplate::class);
    $services->alias(PageTemplateInterface::class, PageTemplate::class);
```
That makes it clear to me that I have a PageTemplate service aliased to PageTemplateInterface.

And which version of the framework are you using?

**Comment by reypm** [comment, marked as answer]
@ahundiak @ro0NL at the very end and after some research this is what I have ended up with:

```php
    
    // dynamically search for submodules with repositories and it's interfaces in the project root
    $finder = (new Finder())
        ->name('*{Repository}.php')
        ->ignoreUnreadableDirs()
        ->ignoreVCS(true)
        ->in(__DIR__.'/../../../')
        ->exclude(['framework', 'vendor', 'Dockerfiles'])
        ->files()
    ;

    foreach ($finder as $file) {
        // some magic to obtain the *Repository and *RepositoryInterface
        $repository = substr(str_replace('/', '\\', $file->getRelativePathname()), 0, -4);
        $interface = substr(str_replace('/', '\\', $file->getRelativePathname()), 0, -4).'Interface';

        if (!class_exists($repository)) {
            throw new Exception(sprintf('The repository %s is not valid.', $repository));
        }

        if (!interface_exists($interface)) {
            throw new Exception(sprintf('The interface %s is not valid.', $interface));
        }

        $reflectedRepository = new ReflectionClass($repository);
        if (!$reflectedRepository->implementsInterface($interface)) {
            throw new Exception(sprintf('Repository %s must implement %s.', $repository, $interface));
        }
        
        // register the service 
        $services->set($repository)->args([service(ManagerRegistry::class)])->alias($interface, $repository);
    }
```

if any of you have some ideas for improvement and/or suggestions I am open to them :) 

thanks both for taking the time to help me
